### ISSUE TITLE
Queue & Deque segfault upon GC

### ISSUE THREAD

**Issue by orls** [php-ds/extension#114]
`Queue` & `Deque` can segfault when GC is triggered:

```
<?php

$q = new \Ds\Queue();

for ($i = 0; $i < 100 ; $i++) {
    $q->push(123);
    gc_collect_cycles();
}
```

Observed on:
- php 7.2.5 & ds 1.2.5, Mac & linux (via docker for mac);
- php 7.1.14 & ds 1.2.3, Mac

Not observed on php 7.1.4 & ds 1.2.3 (using [this docker image](https://hub.docker.com/r/devdrops/php-ds/) mentioned in another issue).

I can avoid this locally by [altering `php_ds_queue_get_gc`](https://github.com/php-ds/extension/blob/v1.2.5/src/php/handlers/php_queue_handlers.c#L55)  as follows (and similarly`php_ds_dueue_get_gc`):

```patch
     ds_queue_t *queue = Z_DS_QUEUE_P(obj);

     *gc_data  = queue->deque->buffer;
-    *gc_count = (int) queue->deque->capacity;
+    *gc_count = (int) queue->deque->size;

     return NULL;
 }
```

...but I'm not confident that this is a proper fix, since the bug seems to indicate that the spare capacity in the buffer is mis-initialized. Perhaps something about the lower-level mem mgmt used by [deque reallocation](https://github.com/php-ds/extension/blob/v1.2.5/src/ds/ds_deque.c#L141) has changed in PHP?

**Comment by rtheunissen** [comment]
> I'm not confident that this is a proper fix, since the bug seems to indicate that the spare capacity in the buffer is mis-initialized.

I think you're absolutely right here. I'm assuming that the GC is stepping across the buffer but encounters a bad zval, and fails. I don't think using `size` is safe here because the values in the buffer aren't necessarily positioned between 0 and size-1 (in the case of deque, at least). So the only solution I see here is to investigate how the buffer is allocated. 

It's also worth noting that Queue uses a Deque internally so fixing Deque's GC will fix Queue as well.

**Comment by orls** [comment]
Hm – upon further investigation, I can't (yet) recreate this on the official docker php image. I'll try to dig deeper to see if it's connected to another extension.

**Comment by rtheunissen** [comment]
I haven't encountered this myself, but I'll try to reproduce this locally today and see what happens. 👍

**Comment by rtheunissen** [comment]
@orls keep in mind that unallocated memory is unpredictable, and this might only segfault _sometimes.

**Comment by rtheunissen** [comment]
I have a feeling we're not clearing out on reallocation. See https://lxr.room11.org/xref/php-src%40master/ext/spl/spl_fixedarray.c#123

**Comment by orls** [comment]
> keep in mind that unallocated memory is unpredictable, and this might only segfault sometimes.

Indeed. Fiddling with extensions does seem to affect it, but I doubt that's related to the particular extensions involved, it'll just be happenstance of memory allocations over time.

Here's the best recreation I can get, by restricting memory & allocating some stuff first. Using latest official php docker image (7.2.5)

```
<?php

ini_set('memory_limit','8M');

$junk = [];
for ($i = 0; $i < 1024; $i++) {
    $junk []= random_bytes(1024);
}

$q = new \Ds\Queue();
for ($i = 0; $i < 100 ; $i++) {
    $q->push(123);
    gc_collect_cycles();
    echo $q->count() . PHP_EOL;
}
```

**Comment by rtheunissen** [comment]
This has now been fixed and released as 1.2.6 on [PECL](https://pecl.php.net/package/ds/1.2.6).
